File: mycollab-web-pro/src/main/java/com/mycollab/pro/module/project/view/finance/ProjectGenericItemTooltipGenerator.java
package com.mycollab.pro.module.project.view.finance;

import com.mycollab.module.project.ProjectTooltipGenerator;
import com.mycollab.module.project.ProjectTypeConstants;
import com.mycollab.module.project.domain.SimpleRisk;
import com.mycollab.module.project.domain.SimpleStandupReport;
import com.mycollab.module.project.domain.SimpleTask;
import com.mycollab.module.project.service.ProjectTaskService;
import com.mycollab.module.project.service.RiskService;
import com.mycollab.module.project.service.StandupReportService;
import com.mycollab.module.tracker.domain.SimpleBug;
import com.mycollab.module.tracker.domain.SimpleComponent;
import com.mycollab.module.tracker.domain.SimpleVersion;
import com.mycollab.module.tracker.service.BugService;
import com.mycollab.module.tracker.service.ComponentService;
import com.mycollab.module.tracker.service.VersionService;
import com.mycollab.spring.AppContextUtil;
import com.mycollab.vaadin.AppUI;
import com.mycollab.vaadin.UserUIContext;

import java.time.ZoneId;

/**
 * @author MyCollab Ltd.
 * @since 4.0.0
 */
class ProjectGenericItemTooltipGenerator {
    private String html;

    ProjectGenericItemTooltipGenerator(String type, int typeId) {
        html = "";
        Integer sAccountId = AppUI.getAccountId();
        ZoneId timeZone = UserUIContext.getUserTimeZone();
        String siteURL = AppUI.getSiteUrl();

        if (ProjectTypeConstants.BUG.equals(type)) {
            BugService service = AppContextUtil.getSpringBean(BugService.class);
            SimpleBug bug = service.findById(typeId, sAccountId);
            html = ProjectTooltipGenerator.generateToolTipBug(UserUIContext.getUserLocale(), AppUI.getDateFormat(),
                    bug, siteURL, timeZone, false);
        } else if (ProjectTypeConstants.TASK.equals(type)) {
            ProjectTaskService service = AppContextUtil.getSpringBean(ProjectTaskService.class);
            SimpleTask task = service.findById(typeId, sAccountId);
            html = ProjectTooltipGenerator.generateToolTipTask(UserUIContext.getUserLocale(), AppUI.getDateFormat(),
                    task, siteURL, timeZone, false);
        } else if (ProjectTypeConstants.RISK.equals(type)) {
            RiskService service = AppContextUtil.getSpringBean(RiskService.class);
            SimpleRisk risk = service.findById(typeId, sAccountId);
            html = ProjectTooltipGenerator.generateToolTipRisk(UserUIContext.getUserLocale(), AppUI.getDateFormat(),
                    risk, siteURL, timeZone, false);
        } else if (ProjectTypeConstants.BUG_VERSION.equals(type)) {
            VersionService service = AppContextUtil.getSpringBean(VersionService.class);
            SimpleVersion version = service.findById(typeId, sAccountId);
            html = ProjectTooltipGenerator.generateToolTipVersion(UserUIContext.getUserLocale(), AppUI.getDateFormat(),
                    version, siteURL, timeZone);
        } else if (ProjectTypeConstants.BUG_COMPONENT.equals(type)) {
            ComponentService service = AppContextUtil.getSpringBean(ComponentService.class);
            SimpleComponent component = service.findById(typeId, sAccountId);
            html = ProjectTooltipGenerator.generateToolTipComponent(UserUIContext.getUserLocale(), component, siteURL, timeZone);
        } else if (ProjectTypeConstants.STANDUP.equals(type)) {
            StandupReportService service = AppContextUtil.getSpringBean(StandupReportService.class);
            SimpleStandupReport standup = service.findById(typeId, sAccountId);
            html = ProjectTooltipGenerator.generateToolTipStandUp(UserUIContext.getUserLocale(), AppUI.getDateFormat(),
                    standup, siteURL, timeZone);
        }
    }

    public String getContent() {
        return html;
    }
}

File: mycollab-web-pro/src/main/java/com/mycollab/pro/module/project/view/finance/ProjectGenericItemTooltipGenerator.kt
package com.mycollab.pro.module.project.view.finance

import com.mycollab.core.UnsupportedFeatureException
import com.mycollab.module.project.ProjectTooltipGenerator
import com.mycollab.module.project.ProjectTypeConstants
import com.mycollab.module.project.domain.SimpleRisk
import com.mycollab.module.project.domain.SimpleStandupReport
import com.mycollab.module.project.domain.SimpleTask
import com.mycollab.module.project.service.TaskService
import com.mycollab.module.project.service.RiskService
import com.mycollab.module.project.service.StandupReportService
import com.mycollab.module.project.domain.SimpleBug
import com.mycollab.module.project.domain.SimpleComponent
import com.mycollab.module.project.domain.SimpleVersion
import com.mycollab.module.project.service.BugService
import com.mycollab.module.project.service.ComponentService
import com.mycollab.module.project.service.VersionService
import com.mycollab.spring.AppContextUtil
import com.mycollab.vaadin.AppUI
import com.mycollab.vaadin.UserUIContext

import java.time.ZoneId

/**
 * @author MyCollab Ltd.
 * @since 4.0.0
 */
internal object ProjectGenericItemTooltipGenerator {

    @JvmStatic
    fun generateTooltipForEntity(type: String, typeId: Int): String? {
        val sAccountId = AppUI.accountId
        val timeZone = UserUIContext.getUserTimeZone()
        val siteURL = AppUI.siteUrl

        when (type) {
            ProjectTypeConstants.BUG -> {
                val service = AppContextUtil.getSpringBean(BugService::class.java)
                val bug = service.findById(typeId, sAccountId)
                return ProjectTooltipGenerator.generateToolTipBug(UserUIContext.getUserLocale(), AppUI.dateFormat,
                        bug, siteURL, timeZone, false)
            }
            ProjectTypeConstants.TASK -> {
                val service = AppContextUtil.getSpringBean(TaskService::class.java)
                val task = service.findById(typeId, sAccountId)
                return ProjectTooltipGenerator.generateToolTipTask(UserUIContext.getUserLocale(), AppUI.dateFormat,
                        task, siteURL, timeZone, false)
            }
            ProjectTypeConstants.RISK -> {
                val service = AppContextUtil.getSpringBean(RiskService::class.java)
                val risk = service.findById(typeId, sAccountId)
                return ProjectTooltipGenerator.generateToolTipRisk(UserUIContext.getUserLocale(), AppUI.dateFormat,
                        risk, siteURL, timeZone, false)
            }
            ProjectTypeConstants.VERSION -> {
                val service = AppContextUtil.getSpringBean(VersionService::class.java)
                val version = service.findById(typeId, sAccountId)
                return ProjectTooltipGenerator.generateToolTipVersion(UserUIContext.getUserLocale(), AppUI.dateFormat,
                        version, siteURL, timeZone)
            }
            ProjectTypeConstants.COMPONENT -> {
                val service = AppContextUtil.getSpringBean(ComponentService::class.java)
                val component = service.findById(typeId, sAccountId)
                return ProjectTooltipGenerator.generateToolTipComponent(UserUIContext.getUserLocale(), component, siteURL, timeZone)
            }
            ProjectTypeConstants.STANDUP -> {
                val service = AppContextUtil.getSpringBean(StandupReportService::class.java)
                val standup = service.findById(typeId, sAccountId)
                return ProjectTooltipGenerator.generateToolTipStandUp(UserUIContext.getUserLocale(), AppUI.dateFormat,
                        standup, siteURL, timeZone)
            }
            else -> throw UnsupportedFeatureException("Not support tooltip with type")
        }
    }
}
